![Imgur](https://i.imgur.com/5pXzCIu.png)

# Data Science va Sun'iy Intellekt Praktikum

## 3-MODUL. MA’LUMOTLARGA ISHLOV BERISH

### 3.2 Ma'lumotlarni tozalash va tayyorlash

## Boshqa foydali funksiyalar

### No-odatiy qiymatlarni topish

In [2]:
import pandas as pd
import numpy as np

In [3]:
df = pd.read_csv("https://github.com/anvarnarz/praktikum_datasets/raw/main/MOCK_USER_DATA.csv", index_col=0)
df.head()

,first_name,last_name,email,gender,age
id,,,,,
1,Dorothy,Verryan,dverryan0@who.int,Non-binary,1
2,Pietrek,Scanlin,pscanlin1@yolasite.com,Male,91
3,Lilith,Sarjeant,lsarjeant2@creativecommons.org,Genderfluid,1233
4,Knox,Cunningham,kcunningham3@yellowpages.com,Genderfluid,6
5,Mohammed,Habbershon,mhabbershon4@sourceforge.net,Female,48


Yuqoridagi jadvlada odamlarning yoshi jamlangan. Keling jadval haqida qisqa ma'lumot olamiz:

In [4]:
df.describe()

,age
count,100.000000
mean,145.660000
std,565.631411
min,-85.000000
25%,22.750000
50%,53.500000
75%,84.250000
max,4444.000000


Nimadur xato: `max` yosh `4444` deb ko'rsatayapti, tabiiyku bunday bo'lishi mumkin emas. `min` da ham manfiy qiymat ko'rishimiz mumkin, bu ham xato. Demak jadvalimizda noto'g'ri qiymatlar ham kirib qolgan.

Keling, boshlanishiga katta qiymatlarni filtrlab olamiz, yaqin tarixda odam yoshi maksimum `123` dan oshmagnini bilgan holda, keling shu qiymatdan katta qatorlarni topamiz:

In [5]:
df[df.age>123]

,first_name,last_name,email,gender,age
id,,,,,
3,Lilith,Sarjeant,lsarjeant2@creativecommons.org,Genderfluid,1233
39,Godard,Miere,gmiere12@1688.com,Polygender,3434
47,Cleve,Sparshatt,csparshatt1a@economist.com,Genderfluid,333
49,Trace,Stockney,tstockney1c@flavors.me,Female,344
65,Eloise,Snoddon,esnoddon1s@elegantthemes.com,Genderfluid,332
80,Gabie,Rothon,grothon27@ehow.com,Genderqueer,4444
100,Mable,Toffano,mtoffano2r@myspace.com,Bigender,233


Qatorlarning indeksini ham ajratib olishimiz mumkin:

In [6]:
idx = df[df.age>123].index
idx

Index([3, 39, 47, 49, 65, 80, 100], dtype='int64', name='id')

Manfiy qiymatlarni ham yuqoridagi usul bilan filtrlab olishimiz mumkin:

In [7]:
df[df.age<0]

,first_name,last_name,email,gender,age
id,,,,,
11,Elijah,Heine,eheinea@imdb.com,Bigender,-85
19,Luisa,Bradborne,lbradbornei@fema.gov,Female,-75
21,Levin,Worner,lwornerk@ihg.com,Genderqueer,-43
33,Ebenezer,Greeves,egreevesw@myspace.com,Male,-26
38,Hana,Langwade,hlangwade11@washingtonpost.com,Genderqueer,-42
90,Zachariah,Meins,zmeins2h@weather.com,Male,-69


Ikki filtrni jamlab yozish ham mumkin:

In [8]:
df[(df.age>123) | (df.age<0)]

,first_name,last_name,email,gender,age
id,,,,,
3,Lilith,Sarjeant,lsarjeant2@creativecommons.org,Genderfluid,1233
11,Elijah,Heine,eheinea@imdb.com,Bigender,-85
19,Luisa,Bradborne,lbradbornei@fema.gov,Female,-75
21,Levin,Worner,lwornerk@ihg.com,Genderqueer,-43
33,Ebenezer,Greeves,egreevesw@myspace.com,Male,-26
38,Hana,Langwade,hlangwade11@washingtonpost.com,Genderqueer,-42
39,Godard,Miere,gmiere12@1688.com,Polygender,3434
47,Cleve,Sparshatt,csparshatt1a@economist.com,Genderfluid,333
49,Trace,Stockney,tstockney1c@flavors.me,Female,344


Qatorlar bilan qanday yo'l tutish sizga bog'liq. Manfiy qiymatlarni musbatga o'tkazishimiz mumkin, buning uchun `numpy` dagi `abs` funksiyasini qo'llaymiz:

In [9]:
df['age'] = df['age'].apply(np.abs)
df.head(15)

,first_name,last_name,email,gender,age
id,,,,,
1,Dorothy,Verryan,dverryan0@who.int,Non-binary,1
2,Pietrek,Scanlin,pscanlin1@yolasite.com,Male,91
3,Lilith,Sarjeant,lsarjeant2@creativecommons.org,Genderfluid,1233
4,Knox,Cunningham,kcunningham3@yellowpages.com,Genderfluid,6
5,Mohammed,Habbershon,mhabbershon4@sourceforge.net,Female,48
6,Glynn,Ebbing,gebbing5@rakuten.co.jp,Genderfluid,77
7,Wendy,Popland,wpopland6@cbc.ca,Genderfluid,67
8,Nye,Anthiftle,nanthiftle7@examiner.com,Agender,99
9,Der,Neames,dneames8@ycombinator.com,Bigender,77


Katta qiymatlarni tashlab yuboramiz:

In [ ]:
idx = df[df.age>123].index
df.drop(index=idx, inplace=True)
df.describe()

,age
count,93.000000
mean,52.612903
std,29.978265
min,1.000000
25%,29.000000
50%,52.000000
75%,77.000000
max,100.000000


Ko'rib turganingizdek, jadval qiymatlari normal holatga qaytdi.

Jadvlada yana bir e'tiborimni tortgna jihati, jinslar faqatgina erkak (`Male`) va ayol (`Female`) emas.

In [11]:
df.gender.value_counts()

,count
gender,
Male,21
Genderfluid,14
Bigender,13
Female,12
Genderqueer,11
Non-binary,10
Polygender,10
Agender,9


Agar bu sizning keyingi hisoblaringizga ta'sir o'tkazadi desangiz, bu qiymatlarni ham filtrlab olishingiz mumkin:

In [12]:
df[df['gender'].isin(['Male','Female'])]

,first_name,last_name,email,gender,age
id,,,,,
2,Pietrek,Scanlin,pscanlin1@yolasite.com,Male,91
5,Mohammed,Habbershon,mhabbershon4@sourceforge.net,Female,48
10,David,Meredyth,dmeredyth9@telegraph.co.uk,Female,68
12,Raffaello,Firminger,rfirmingerb@bloomberg.com,Female,74
17,Byram,Martynikhin,bmartynikhing@jiathis.com,Male,40
19,Luisa,Bradborne,lbradbornei@fema.gov,Female,75
22,Ray,Padly,rpadlyl@twitpic.com,Male,85
26,Ephrayim,Westmacott,ewestmacottp@ning.com,Male,77
29,Biron,Acklands,backlandss@comsenz.com,Male,13


### Jadvlani aralashtirib tashlash

Ba'zida, tadqiqod uchun jadvaldagi qiymatlar o'rnini aralashtirib tashlash talab qilinishi mumkin. Misol uchun, sizda ma'lumotlar ma'lum bir tartibda turibdi, lekin siz ma'lumotlarni tasodifiy tartibda chiqarmoqchiisiz, buning uchun `numpy` kutubxonasining `random` modulidagi `permutation` funksiyasiga murojaat qilamiz.

Bizda quyidagi jadval bor:

In [13]:
df.head()

,first_name,last_name,email,gender,age
id,,,,,
1,Dorothy,Verryan,dverryan0@who.int,Non-binary,1
2,Pietrek,Scanlin,pscanlin1@yolasite.com,Male,91
3,Lilith,Sarjeant,lsarjeant2@creativecommons.org,Genderfluid,1233
4,Knox,Cunningham,kcunningham3@yellowpages.com,Genderfluid,6
5,Mohammed,Habbershon,mhabbershon4@sourceforge.net,Female,48


Keling, jadval uzunligini topaylik:

In [15]:
len(df)

100

Jadvalimizda 100 ta qator mavjud, biz 1-100 gacha sonlarni tasodifiy tartibda saqlovchi ro'yxat yaratamiz:

In [17]:
randlist = np.random.permutation(100)
randlist

array([58, 49, 44, 41, 37, 60, 54, 91, 33, 74, 13, 29, 20, 72, 88, 21, 80,
       23,  5, 63, 71, 16, 51, 96, 76,  7, 70, 22, 95, 17, 31, 64, 34, 43,
       50, 67, 99, 42, 32, 89, 25, 40, 87, 15, 73, 79, 11, 35, 78, 62, 84,
       94,  9, 27, 90, 66,  2, 38, 39, 69, 28, 82, 30, 92, 14, 81, 36, 12,
       19, 85, 57, 83,  4, 48, 55, 93,  8,  0, 65, 75, 10, 47, 61, 26, 52,
        3, 46, 56, 45, 18, 24, 77, 59, 97,  6,  1, 53, 98, 68, 86])

Navbat, `df` ni yuqoridagi tasodifiy tartibda tahlash, buning uchun `.take()` meotidga murojaat qilamiz

In [18]:
df.take(randlist)

,first_name,last_name,email,gender,age
id,,,,,
59,Lynna,Gionettitti,lgionettitti1m@unc.edu,Polygender,70
50,Bartolomeo,Heffron,bheffron1d@bluehost.com,Agender,9
45,Koralle,Pumphreys,kpumphreys18@issuu.com,Polygender,38
42,Dermot,Dragge,ddragge15@google.co.jp,Genderfluid,37
38,Hana,Langwade,hlangwade11@washingtonpost.com,Genderqueer,42
...,...,...,...,...,...
2,Pietrek,Scanlin,pscanlin1@yolasite.com,Male,91
54,Gregory,Criag,gcriag1h@purevolume.com,Polygender,50
99,Desdemona,Jurasek,djurasek2q@aboutads.info,Genderfluid,46


Marhamat, jadvalimiz tasodifiy tartibda tahlandi.

### Jadvladan  tasodifiy qiymatlarni chiqarib olish.
Buning uchun bizga avvaldan tanish .sample() metodiga murojaat qilamiz va aynan nechta qiymat kerak ekanini ham ko'rsatamiz.

Bunda tasodifiy qatorlar **takrorlanmagan** holda chiqadi.

In [19]:
df.sample(5)

,first_name,last_name,email,gender,age
id,,,,,
27,Natalee,Baudone,nbaudoneq@marriott.com,Genderqueer,74
56,Guillermo,Batrick,gbatrick1j@macromedia.com,Bigender,29
18,Artemus,Shovelin,ashovelinh@sphinn.com,Non-binary,58
11,Elijah,Heine,eheinea@imdb.com,Bigender,85
66,Alejoa,Abrahamoff,aabrahamoff1t@com.com,Polygender,28


Agar qatorlar takrorlanishini istasangiz, `replace=True` parametrini ko'rsatasiz:

In [23]:
df.sample(10, replace=True)

,first_name,last_name,email,gender,age
id,,,,,
11,Elijah,Heine,eheinea@imdb.com,Bigender,85
22,Ray,Padly,rpadlyl@twitpic.com,Male,85
84,Tildi,Phythien,tphythien2b@ustream.tv,Genderfluid,76
68,Letty,Liddell,lliddell1v@nps.gov,Genderfluid,85
68,Letty,Liddell,lliddell1v@nps.gov,Genderfluid,85
78,Richy,Cornilleau,rcornilleau25@nhs.uk,Bigender,88
71,Zane,Brickham,zbrickham1y@soundcloud.com,Genderqueer,5
38,Hana,Langwade,hlangwade11@washingtonpost.com,Genderqueer,42
39,Godard,Miere,gmiere12@1688.com,Polygender,3434


Ahamiyat bering, `68` indeks ikki marta chiqdi.